--- /web/pernmband/tmp/netserver.c	Sat Sep 23 15:52:21 2000
+++ netserver.c	Sat Sep 23 15:44:12 2000
@@ -169,6 +169,7 @@
 	playing_receive[PKT_DESTROY]		= Receive_destroy;
 	playing_receive[PKT_LOOK]		= Receive_look;
 	playing_receive[PKT_SPELL]		= Receive_spell;
+	playing_receive[PKT_FIGHT]		= Receive_fight;
 
 	playing_receive[PKT_OPEN]		= Receive_open;
 	playing_receive[PKT_PRAY]		= Receive_pray;
@@ -219,6 +220,8 @@
 	playing_receive[PKT_MASTER]		= Receive_master;
 
 	playing_receive[PKT_AUTOPHASE]		= Receive_autophase;
+
+	playing_receive[PKT_CLEAR_BUFFER]	= Receive_clear_buffer;
 }
 
 static int Init_setup(void)
@@ -279,9 +282,9 @@
 
 		/* Get our hostname */
 #ifdef BIND_NAME
-		strncpy( local_name, BIND_NAME, 1024);
+			strncpy(local_name, BIND_NAME, 1024);
 #else
-		GetLocalHostName(local_name, 1024);
+			GetLocalHostName(local_name, 1024);
 #endif
 	}
 
@@ -734,7 +737,7 @@
 		return status;
 	}
 
-	if (version < 0x0420)
+	if (version < 0x0000)
 	{
 		return E_VERSION;
 	}
@@ -1328,6 +1331,15 @@
 			return -1;
 		}
 	}
+	
+	/* Read class extra */	
+//	n = Packet_scanf(&connp->r, "%hd", &connp->class_extra);
+
+	if (n <= 0)
+	{
+		Destroy_connection(ind, "Misread class extra");
+		return -1;
+	}
 
 	/* Read the options */
 	for (i = 0; i < 64; i++)
@@ -3637,7 +3649,18 @@
 
 	if (connp->id != -1 && p_ptr->energy >= level_speed(p_ptr->dun_depth))
 	{
-		do_cmd_cast(player, book, spell);
+		if (p_ptr->pclass == CLASS_SORCERER)
+		{
+			do_cmd_sorc(player, book, spell);
+		}
+		else if (p_ptr->pclass == CLASS_ROGUE)
+		{
+			do_cmd_shad(player, book, spell);
+		}
+		else
+		{
+			do_cmd_cast(player, book, spell);
+		}
 		return 2;
 	}
 	else if (player)
@@ -3725,6 +3748,45 @@
 	return 1;
 }
 
+static int Receive_fight(int ind)
+{
+	connection_t *connp = &Conn[ind];
+	player_type *p_ptr;
+
+	char ch;
+
+	int n, player;
+
+	s16b book, technic;
+
+	if (connp->id != -1)
+	{
+		player = GetInd[connp->id];
+		p_ptr = Players[player];
+	}
+
+	if ((n = Packet_scanf(&connp->r, "%c%hd%hd", &ch, &book, &technic)) <= 0)
+	{
+		if (n == -1)
+			Destroy_connection(ind, "read error");
+		return n;
+	}
+
+	if (connp->id != -1 && p_ptr->energy >= level_speed(p_ptr->dun_depth))
+	{
+		do_cmd_fight(player, book, technic);
+		return 2;
+	}
+	else if (player)
+	{
+		p_ptr->current_spell = -1;
+		Packet_printf(&connp->q, "%c%hd%hd", ch, book, technic);
+		return 0;
+	}
+
+	return 1;
+}
+
 static int Receive_ghost(int ind)
 {
 	connection_t *connp = &Conn[ind];
@@ -4922,6 +4984,36 @@
 	return 1;
 }
 
+static int Receive_clear_buffer(int ind)
+{
+	connection_t *connp = &Conn[ind];
+	player_type *p_ptr;
+	int player, n;
+	char ch;
+
+	if (connp->id != -1)
+	{
+		player = GetInd[connp->id];
+		p_ptr = Players[player];
+	}
+	else player = 0;
+
+	if ((n = Packet_scanf(&connp->r, "%c", &ch)) <= 0)
+	{
+		if (n == -1)	
+			Destroy_connection(ind, "read error");
+		return n;
+	}
+
+	if (player)
+	{
+		/* Clear the buffer */
+		Sockbuf_clear(&connp->q);
+	}
+
+	return 1;
+}
+
 static int Receive_special_line(int ind)
 {
 	connection_t *connp = &Conn[ind];
@@ -5118,8 +5210,14 @@
 			do_cmd_ghost_power_aux(Ind, dir);
 		else if (p_ptr->mp_ptr->spell_book == TV_MAGIC_BOOK)
 			do_cmd_cast_aux(Ind, dir);
+		else if (p_ptr->mp_ptr->spell_book == TV_SORCERY_BOOK)
+			do_cmd_sorc_aux(Ind, dir);
 		else if (p_ptr->mp_ptr->spell_book == TV_PRAYER_BOOK)
 			do_cmd_pray_aux(Ind, dir);
+		else if (p_ptr->mp_ptr->spell_book == TV_FIGHT_BOOK)
+			do_cmd_fight_aux(Ind, dir);
+		else if (p_ptr->mp_ptr->spell_book == TV_SHADOW_BOOK)
+			do_cmd_shad_aux(Ind, dir);
 		else p_ptr->current_spell = -1;
 	}
 	else if (p_ptr->current_rod != -1)
