*** spells2.c~	Fri Mar 27 20:19:00 1998
--- spells2.c	Tue Apr  7 11:50:22 1998
***************
*** 799,813 ****
                  }
                  if (p_ptr->muta1 & MUT1_MIND_BLST)
                  {
!                     info[i++] = " You can Mind Blast your enemies.\n";
                  }
                  if (p_ptr->muta1 & MUT1_RADIATION)
                  {
!                     info[i++] = " You can emit hard radiation at will.\n";
                  }
                  if (p_ptr->muta1 & MUT1_VAMPIRISM)
                  {
!                     info[i++] = " You can drain life from a foe like a vampire.\n";
                  }
                  
          }
--- 799,813 ----
                  }
                  if (p_ptr->muta1 & MUT1_MIND_BLST)
                  {
!                     info[i++] = " You can Mind Blast your enemies.";
                  }
                  if (p_ptr->muta1 & MUT1_RADIATION)
                  {
!                     info[i++] = " You can emit hard radiation at will.";
                  }
                  if (p_ptr->muta1 & MUT1_VAMPIRISM)
                  {
!                     info[i++] = " You can drain life from a foe like a vampire.";
                  }
                  
          }
